### PR TITLE
[bugfix] Resolve memory leak for evaluation

### LINKED ISSUE #5735: Bounded memory leak caused by `trainer.evalutaion_loop.outputs`
## 🐛 Bug

`trainer.evalutaion_loop.outputs` caches the outputs of every validation steps in `def run_evaluation(self, max_batches=None):` of trainer:
https://github.com/PyTorchLightning/pytorch-lightning/blob/d71659b42a13946b854d49f5bb1bf6e2bcd5b9b2/pytorch_lightning/trainer/trainer.py#L659

It's not reset until the start of the next validation epoch:

https://github.com/PyTorchLightning/pytorch-lightning/blob/d71659b42a13946b854d49f5bb1bf6e2bcd5b9b2/pytorch_lightning/trainer/trainer.py#L621

https://github.com/PyTorchLightning/pytorch-lightning/blob/d71659b42a13946b854d49f5bb1bf6e2bcd5b9b2/pytorch_lightning/trainer/evaluation_loop.py#L124-L128

### To Reproduce
Sorry, my working environment forbids me to use google drive.

```python
import torch
from pytorch_lightning import LightningModule
from torch.utils.data import Dataset

class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def loss(self, batch, prediction):
        # An arbitrary loss to have a loss that updates the model weights during `Trainer.fit` calls
        return torch.nn.functional.mse_loss(prediction, torch.ones_like(prediction))
    
    def on_train_epoch_start(self):
        print('Before delete:', torch.cuda.memory_allocated())
        for out in self.trainer.evaluation_loop.outputs[0]:
          if 'x' in out:
            del out['x']
        print('After delete:', torch.cuda.memory_allocated())

    def training_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"loss": loss}

    def training_step_end(self, training_step_outputs):
        return training_step_outputs

    def training_epoch_end(self, outputs) -> None:
        torch.stack([x["loss"] for x in outputs]).mean()

    def validation_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"x": loss}

    def validation_epoch_end(self, outputs) -> None:
        torch.stack([x['x'] for x in outputs]).mean()

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
        return [optimizer], [lr_scheduler]
```
```python
def test_x(tmpdir):
    # init model
    model = BoringModel()

    # Initialize a trainer
    trainer = pl.Trainer(
        max_epochs=1, 
        progress_bar_refresh_rate=20,
        gpus=[0],
    )

    # Train the model ⚡
    trainer.fit(model, train, val)
```
#### Execution
```python
test_x(tmpdir)
```
#### Output
```bash
GPU available: True, used: True
TPU available: None, using: 0 TPU cores
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]

  | Name  | Type   | Params
---------------------------------
0 | layer | Linear | 66    
---------------------------------
66        Trainable params
0         Non-trainable params
66        Total params
Validation sanity check: 0%
0/2 [00:00<?, ?it/s]
Epoch 0: 100%
626/626 [00:00<00:00, 653.87it/s, loss=2.5e-14, v_num=6]
Before delete: 2048
After delete: 1024
```

### Expected behavior
There shouldn't be such cached tensors.
This may cause OOM in some cases that OOM can be avoided.
For example,
- On the first training epoch, a model fitted just right in the GPU memory runs fine w/o OOM.
- After the first validation epoch, some GPU tensors are retained and occupy some portion of the memory.
- On the second training epoch, the same model encounters OOM error.

Clear all references to those validation output tensors at the end of the validation epoch.
Maybe, more specifically, at here?
https://github.com/PyTorchLightning/pytorch-lightning/blob/d71659b42a13946b854d49f5bb1bf6e2bcd5b9b2/pytorch_lightning/trainer/evaluation_loop.py#L224-L229

### Environment

* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     True
	- pyTorch_version:   1.7.0+cu101
	- pytorch-lightning: 1.1.6
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.9
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR delete self.trainer.evaluation_loop.outputs on `validation_epoch_end` call.

Fixes #5735

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by kaushikb11 — APPROVED** [review]

**Review by carmocca — CHANGES_REQUESTED** [review]
Please keep the `logging_` directory name as that's done on purpose to not collide with the stdlib `logging`

File: tests/trainer/logging_/test_eval_loop_logging_1_0.py [hunk 1]
```diff
@@ -110,6 +110,7 @@ def training_step(self, batch, batch_idx):
             return acc
 
         def validation_step(self, batch, batch_idx):
+            self.batch_idx = batch_idx
```
Comment (carmocca) [inline tests/trainer/logging_/test_eval_loop_logging_1_0.py:113]: Any reason to add this?

**Review by carmocca — APPROVED** [review]

File: tests/trainer/logging_/test_eval_loop_logging_1_0.py (hunk 1 above)
Comment (tchaton) [reply tests/trainer/logging_/test_eval_loop_logging_1_0.py:113]: removed !

**Review by carmocca — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]
